controller/registries/consul/config.go

```diff
@@ -181,46 +258,145 @@ func (reg *Consul) fetchAllServices(client *Client) (map[consulService]bool, err
+			Protocol: string(protocol),
+		}
+		if len(portList) == 0 {
+			portList = append(portList, port)
+		}
+
+		endpoint := istioapi.WorkloadEntry{
+			Address: service.Service.Address,
+			Ports:   map[string]uint32{port.Protocol: port.Number},
+			Labels:  service.Service.Meta,
+		}
+		endpoints = append(endpoints, &endpoint)
+	}
+
+	return &registry.ServiceEntryWrapper{
+		ServiceEntry: istioapi.ServiceEntry{
+			Hosts:      []string{host},
+			Ports:      portList,
+			Location:   istioapi.ServiceEntry_MESH_INTERNAL,
+			Resolution: istioapi.ServiceEntry_STATIC,
+			Endpoints:  endpoints,
+		},
+		Source: RegistryType,
+	}
+}
+
+func (reg *Consul) subscribe(tag, serviceName string) error {
+	plan, err := watch.Parse(map[string]interface{}{
+		"type":    "service",
+		"service": serviceName,
+	})
+	if err != nil {
+		return err
+	}
+
+	plan.Handler = reg.getSubscribeCallback(tag, serviceName)
+	plan.Token = reg.client.Token
+	plan.Datacenter = reg.client.DataCenter
+	reg.subscriptions[serviceName] = plan
+
+	go func() {
+		err := plan.Run(reg.client.Address)
+		if err != nil {
+			reg.logger.Errorf("failed to subscribe ,err=%v", err)
+		}
+	}()
+
 	return nil
 }
 
+func (reg *Consul) getSubscribeCallback(tag, serviceName string) func(idx uint64, data interface{}) {
+	host := reg.getServiceEntryKey(tag, serviceName)
+	return func(idx uint64, data interface{}) {
+		services, ok := data.([]*consulapi.ServiceEntry)
+		if !ok {
+			reg.logger.Infof("Unexpected type for data in callback: %t", data)
+			return
+		}
+		if reg.stopped.Load() {
+			return
+		}
+		reg.store.Update(host, reg.generateServiceEntry(host, services))
+	}
+
+}
+
 func (reg *Consul) unsubscribe(serviceName string) error {
-	fmt.Println(serviceName)
+	plan, exists := reg.subscriptions[serviceName]
+	if !exists {
+		return fmt.Errorf("no subscription found for service %s", serviceName)
+	}
+
+	plan.Stop()
+	delete(reg.subscriptions, serviceName)
 	return nil
 }
 
 func (reg *Consul) refresh(services map[string][]string) {
+
 	serviceMap := make(map[consulService]bool)
+
 	for serviceName, tags := range services {
-		for _, tag := range tags {
-			service := consulService{
-				Tag:         tag,
-				ServiceName: serviceName,
-			}
-			serviceMap[service] = true
-			if _, ok := reg.watchingServices[service]; !ok {
-				err := reg.subscribe(serviceName)
-				if err != nil {
-					reg.logger.Errorf("failed to subscribe service, err: %v, service: %v", err, serviceName)
-					delete(serviceMap, service)
-				}
+		tag := strings.Join(tags, "-")
+		service := consulService{
+			Tag:         tag,
+			ServiceName: serviceName,
+		}
+		serviceMap[service] = true
+		if _, ok := reg.watchingServices[service]; !ok {
+			err := reg.subscribe("", service.ServiceName)
+			if err != nil {
+				reg.logger.Errorf("failed to subscribe service, err: %v, service: %v", err, service.ServiceName)
+				delete(serviceMap, service)
+				delete(reg.watchingServices, service)
```

Comment:
We don't need to delete service from reg.watchingServices if _, ok := reg.watchingServices[service]; !ok